_drafts/firmware-update-architecture.md

```diff
@@ -0,0 +1,834 @@
+    A -> B [label = "Ld, Updt", fontsize=8];
+
+    E [label = "Updater"];
+    A -> E [label = "Ld, Updt", fontsize=8];
+    E -> A [label = "Updates", fontsize=8];
+    C -> E [label = "Loads", style=dashed, fontsize=8];
+
+    group {
+        label = "Slot 0";
+        color = "PaleGreen";
+        C;
+    }
+    group {
+        label = "Slot 1";
+        color = "LightPink";
+        A;
+    }
+    group {
+        label = "Slot 2";
+        color = "LemonChiffon";
+        B; E;
+    }
+}
+{% endblockdiag %}{:.diag4}
+
+## Implementation Cookbook
+
+I have put together a full implementation of the Bootloader, the Loader, and the
+Application in the [Interrupt
+Github
+repository](https://github.com/memfault/interrupt/tree/master/example/fwup-architecture).
+While discussing every line in details is outside of the scope of this
+conversation, I want to highlight a few patterns I have learned over the years.
+These include ways to package firmwar images, write them to flash, share data
+between programs, and more!
+
+This posts builds upon many ideas previously written about on Interrupt. If you
+haven't read them already, I recommend the following:
+* [How to Write a Bootloader from Scratch]({% post_url
+  2019-08-13-how-to-write-a-bootloader-from-scratch %})
+* [How to Write Linker Scripts for Firmware]({% post_url
+  2019-06-25-how-to-write-linker-scripts-for-firmware %})
+* [GNU Build IDs for Firmwre]({% post_url 2019-05-29-gnu-build-id-for-firmware
+  %})
+* [Building a Tiny CLI Shell for Tiny Firmware]({% post_url
+  2020-06-09-firmware-shell %})
+
+### Image Metadata
+
+Firmware images typically need to be bundled alongside some metadata. It
+identifies what type of image it is, how large it is, what version it is, ...
+etc.
+
+That metadata must be:
+1. Easy to find and parse for an external program
+2. Accessible from within the firmware itself
+
+My favorite approach is to add a struct at the beginning of the image to serve
+as a header. Here's what it looks like:
+
+```c
+// image.h
+typedef struct __attribute__((packed)) {
+    uint16_t image_magic;
+    uint16_t image_hdr_version;
+    uint32_t crc;
+    uint32_t data_size;
+    uint8_t image_type;
+    uint8_t version_major;
+    uint8_t version_minor;
+    uint8_t version_patch;
+    uint32_t vector_addr;
+    uint32_t reserved;
+    char git_sha[8];
+} image_hdr_t;
+```
+
+The contents of that header will vary per application, but you must include:
+1. A way to indicate that the header is valid (here we used a constant in
+`image_magic`)
+2. A version for the header, and
+3. The address of the start of your vector table.
+
+Each of our images must fill in its own header. For example, here it is for my
+Application:
+
+```c
+// app.c
+image_hdr_t image_hdr __attribute__((section(".image_hdr"))) = {
+    .image_magic = IMAGE_MAGIC,
+    .image_hdr_version = IMAGE_VERSION_CURRENT,
+    .image_type = IMAGE_TYPE_APP,
+    .version_major = 1,
```

Comment:
maybe worth explicitly adding the other fields and mentioning they get populated as part of the build process

```
// populated as part of a post compilation step
.crc = 0,
.data_size = 0,
```